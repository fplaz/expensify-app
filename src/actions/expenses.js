// import { v4 as uuidv4 } from 'uuid'
import db from '../firebase/firebase'
import { ref, set, push, child, get, remove, update } from 'firebase/database';

// ADD EXPENSE
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
})

export const startAddExpense = (expenseData = {}) => {
    return async (dispatch, getState) => {
        const uid = getState().auth.uid
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData
        
        const expense = {description, note, amount, createdAt}
        
        const expenseRef = ref(db, 'users/'+ uid + '/expenses')
        const newExpenseRef = push(expenseRef)
        
        try {
            await set(newExpenseRef, expense)
            return dispatch(addExpense({
                id: newExpenseRef.key,
                ...expense
            }))
        } catch(e) {
            console.log(e)
        }
    }
}

//EDIT EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
})

export const StartEditExpense = (id, updates) => {
    return async (dispatc, getStateh) => {
        const uid = getState().auth.uid
        try {
            const dbRef = ref(db, 'users/' + uid + '/expenses/' + id)
            const expense = await update(dbRef, updates)
            return dispatch(editExpense(id, updates))
        }
        catch {
            console.log('StartEditExpense error:' + e)
        }
    }
}

// SET_EXPENSES
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
})

export const startSetExpenses = () => {
    return async (dispatch, getState) => {
        const uid = getState().auth.uid
        try {
            const expenses = []
            const dbRef = ref(db, 'users/' + uid + '/expenses')
            const snapshot = await get(dbRef)
            snapshot.forEach((item) => {
                expenses.push({
                    id: item.key,
                    ...item.val()
                })
            })
            return dispatch(setExpenses(expenses))
        } catch(e) {
            console.log('StartSetExpenses error:' + e)
        }
    }
}

// REMOVE EXPENSE
export const removeExpense = ({id} = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
})

export const startRemoveExpense = (id) => {
    return async (dispatch, getState) => {
        
        const uid = getState().auth.uid
        const expenseRef = ref(db, '/users/' + uid + '/expenses/' + id)

        try {
            await set(expenseRef, null)
            return await dispatch(removeExpense({id}))
        } catch(e) {
            console.log(e)
        }
    }
}