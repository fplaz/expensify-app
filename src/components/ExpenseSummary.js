import React from 'react'
import expensesTotal from '../selectors/expenses-total'
import { connect } from 'react-redux'
import selectExpenses from '../selectors/expenses'

const ExpenseSummary = (props) => (
    <div>
        <p>Viewing {props.expenses.length} expenses totalling {expensesTotal(props.expenses)}</p>
    </div>
)

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
}

export default connect(mapStateToProps)(ExpenseSummary)