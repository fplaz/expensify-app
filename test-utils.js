import React from 'react'
import { render as rtlRender} from '@testing-library/react'
import configureStore from './src/store/configureStore' 
import { Provider } from 'react-redux'

function render(
    ui,
    {
      initialState = {},
      store = configureStore,
      ...renderOptions
    } = {}
  ) {
    const Wrapper = ({ children }) => {
      return <Provider store={store()}>{children}</Provider>;
    }
    return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
  }
  
  // re-export everything
  export * from '@testing-library/react';
  
  // override render method
  export { render };