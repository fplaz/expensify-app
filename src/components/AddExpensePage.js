import React from 'react'
import ExpenseForm from './ExpenseForm'
import { connect } from 'react-redux'
import { startAddExpense } from '../actions/expenses'
import { useNavigate } from 'react-router-dom'

const AddExpensePage = ({ onSubmit }) => {
    const navigate = useNavigate();
  
    const handleOnSubmit = (expense) => {
      onSubmit(expense);
      navigate('/dashboard');
    };
  
    return (
      <div>
        <h1>Add Expense</h1>
        <ExpenseForm onSubmit={handleOnSubmit} />
      </div>
    );
  };
  
  const mapDispatchToProps = (dispatch) => ({
    onSubmit: (expense) => {
      dispatch(startAddExpense(expense));
    },
  });
  
  export default connect(undefined, mapDispatchToProps)(AddExpensePage);