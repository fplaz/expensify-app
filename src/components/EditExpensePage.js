import React from 'react'
import { connect } from 'react-redux'
import ExpenseForm from './ExpenseForm'
import { useNavigate, useParams } from 'react-router-dom'
import { editExpense, startRemoveExpense, removeExpense, StartEditExpense } from '../actions/expenses'

const EditExpensePage = (props) => {
    const navigate = useNavigate()
    return (
        <div>
            <ExpenseForm
                expense={props.expense}
                buttonText="Edit Expense!"
                onSubmit={(expense) => {
                    props.dispatch(StartEditExpense(props.expense.id, expense))
                    navigate('/')
                }}
            />
            <button onClick={() => {
                props.dispatch(startRemoveExpense(props.expense.id))
                navigate('/')
            }}>Remove</button>
        </div>
    )
}

const mapStateToProps = (state) => {
    const params = {id: window.location.pathname.split("/")[2]}
    return {
        expense: state.expenses.find((expense) => {
            return expense.id === params.id
        })
    }
}

export default connect(mapStateToProps)(EditExpensePage)