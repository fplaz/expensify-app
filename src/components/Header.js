import React from 'react'
import {NavLink} from 'react-router-dom'
import { connect } from 'react-redux'
import { startLogout } from '../actions/auth'

export const Header = (props) => (
    <header>
        <h1>Expensify App</h1>
        <NavLink to="/" className={({isActive}) => isActive ? "is-active" : undefined }>Home</NavLink>
        <NavLink to="/create" className={({isActive}) => isActive ? "is-active" : undefined }>Create expense</NavLink>
        <NavLink to="/help" className={({isActive}) => isActive ? "is-active" : undefined }>Help expense</NavLink>
        <button onClick={props.startLogout}>Logout</button>
    </header>
)

const mapDispatchToProps = (dispatch) => ({
    startLogout: () => dispatch(startLogout())
})

export default connect(undefined, mapDispatchToProps)(Header)