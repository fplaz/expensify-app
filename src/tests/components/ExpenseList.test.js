import React from "react"
import { ExpenseList } from '../../components/ExpenseList'
import { MemoryRouter } from 'react-router-dom'
import expenses from "../fixtures/expenses"
import {render} from '../../../test-utils'

test('Should render ExpenseList without expenses', () => {
    const {asFragment} = render(
        <ExpenseList expenses={[]} />
    )
    expect(asFragment()).toMatchSnapshot()
})

test('Should render ExpenseList with expenses', () => {
    const {asFragment} = render(
        <MemoryRouter>
            <ExpenseList expenses={expenses} />
        </MemoryRouter>
    )
    expect(asFragment()).toMatchSnapshot()
})