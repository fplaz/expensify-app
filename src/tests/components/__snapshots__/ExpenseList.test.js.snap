// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpenseList with expenses 1`] = `
<DocumentFragment>
  <div>
    <h1>
      Expense List
    </h1>
    <div>
      <a
        href="/edit/undefined"
      >
        <h3>
          Gum
        </h3>
      </a>
      <p>
        $0.20-January 1st, 2000
      </p>
    </div>
    <div>
      <a
        href="/edit/undefined"
      >
        <h3>
          Rent
        </h3>
      </a>
      <p>
        $109.50-Invalid date
      </p>
    </div>
    <div>
      <a
        href="/edit/undefined"
      >
        <h3>
          Credit Card
        </h3>
      </a>
      <p>
        $4.50-Invalid date
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Should render ExpenseList without expenses 1`] = `
<DocumentFragment>
  <div>
    <h1>
      Expense List
    </h1>
  </div>
</DocumentFragment>
`;
